### PR TITLE
Feat: Add a step to to Request Physical Card form that collects a magic code

### PR DESCRIPTION
**Pull request by hungvu193**
### Details

This PR added a validate code modal when user want to issue a physical card.
### Fixed Issues

$ https://github.com/Expensify/App/issues/50967
PROPOSAL: N/A

### Tests

**Ask @NikkiWines to do the steps for you or follow [this slack guide](https://expensify.slack.com/archives/C03TQ48KC/p1732283244364989) and ensure you request to cancel the card before it ships by asking in Slack**

1. Log into an account with a 
1. Go to your Settings => Wallet.
2. Click on your Physical card => Issue card.
3. Enter all necessary information => Press `Ship card`.
4. Verify that there's modal that allow you to enter magic code.
6. Enter the magic code that's been sent to your login.
7. Verify that you can continue with the flow after that.

- [x] Verify that no errors appear in the JS console

### Offline tests

N/A
### QA Steps

Same as Tests.
- [x] Verify that no errors appear in the JS console

### PR Author Checklist

- [x] I linked the correct issue in the `### Fixed Issues` section above
- [x] I wrote clear testing steps that cover the changes made in this PR
    - [x] I added steps for local testing in the `Tests` section
    - [x] I added steps for the expected offline behavior in the `Offline steps` section
    - [x] I added steps for Staging and/or Production testing in the `QA steps` section
    - [x] I added steps to cover failure scenarios (i.e. verify an input displays the correct error message if the entered data is not correct)
    - [x] I turned off my network connection and tested it while offline to ensure it matches the expected behavior (i.e. verify the default avatar icon is displayed if app is offline)
    - [x] I tested this PR with a [High Traffic account](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#high-traffic-accounts) against the staging or production API to ensure there are no regressions (e.g. long loading states that impact usability).
- [x] I included screenshots or videos for tests on [all platforms](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#make-sure-you-can-test-on-all-platforms)
- [x] I ran the tests on **all platforms** & verified they passed on:
    - [x] Android: Native
    - [x] Android: mWeb Chrome
    - [x] iOS: Native
    - [x] iOS: mWeb Safari
    - [x] MacOS: Chrome / Safari
    - [x] MacOS: Desktop
- [x] I verified there are no console errors (if there's a console error not related to the PR, report it or open an issue for it to be fixed)
- [x] I followed proper code patterns (see [Reviewing the code](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md#reviewing-the-code))
    - [x] I verified that any callback methods that were added or modified are named for what the method does and never what callback they handle (i.e. `toggleReport` and not `onIconClick`)
    - [x] I verified that the left part of a conditional rendering a React component is a boolean and NOT a string, e.g. `myBool && <MyComponent />`.
    - [x] I verified that comments were added to code that is not self explanatory
    - [x] I verified that any new or modified comments were clear, correct English, and explained "why" the code was doing something instead of only explaining "what" the code was doing.
    - [x] I verified any copy / text shown in the product is localized by adding it to `src/languages/*` files and using the [translation method](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60)
      - [x] If any non-english text was added/modified, I verified the translation was requested/reviewed in #expensify-open-source and it was approved by an internal Expensify engineer. Link to Slack message:
    - [x] I verified all numbers, amounts, dates and phone numbers shown in the product are using the [localization methods](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60-L68)
    - [x] I verified any copy / text that was added to the app is grammatically correct in English. It adheres to proper capitalization guidelines (note: only the first word of header/labels should be capitalized), and is either coming verbatim from figma or has been approved by marketing (in order to get marketing approval, ask the Bug Zero team member to add the Waiting for copy label to the issue)
    - [x] I verified proper file naming conventions were followed for any new files or renamed files. All non-platform specific files are named after what they export and are not named "index.js". All platform-specific files are named for the platform the code supports as outlined in the README.
    - [x] I verified the JSDocs style guidelines (in [`STYLE.md`](https://github.com/Expensify/App/blob/main/contributingGuides/STYLE.md#jsdocs)) were followed
- [x] If a new code pattern is added I verified it was agreed to be used by multiple Expensify engineers
- [x] I followed the guidelines as stated in the [Review Guidelines](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md)
- [x] I tested other components that can be impacted by my changes (i.e. if the PR modifies a shared library or component like `Avatar`, I verified the components using `Avatar` are working as expected)
- [x] I verified all code is DRY (the PR doesn't include any logic written more than once, with the exception of tests)
- [x] I verified any variables that can be defined as constants (ie. in CONST.js or at the top of the file that uses the constant) are defined as such
- [x] I verified that if a function's arguments changed that all usages have also been updated correctly
- [x] If any new file was added I verified that:
    - [x] The file has a description of what it does and/or why is needed at the top of the file if the code is not self explanatory
- [x] If a new CSS style is added I verified that:
    - [x] A similar style doesn't already exist
    - [x] The style can't be created with an existing [StyleUtils](https://github.com/Expensify/App/blob/main/src/styles/utils/index.ts) function (i.e. `StyleUtils.getBackgroundAndBorderStyle(theme.componentBG)`)
- [x] If the PR modifies code that runs when editing or sending messages, I tested and verified there is no unexpected behavior for all supported markdown - URLs, single line code, code blocks, quotes, headings, bold, strikethrough, and italic.
- [x] If the PR modifies a generic component, I tested and verified that those changes do not break usages of that component in the rest of the App (i.e. if a shared library or component like `Avatar` is modified, I verified that `Avatar` is working as expected in all cases)
- [x] If the PR modifies a component related to any of the existing Storybook stories, I tested and verified all stories for that component are still working as expected.
- [x] If the PR modifies a component or page that can be accessed by a direct deeplink, I verified that the code functions as expected when the deeplink is used - from a logged in and logged out account.
- [x] If the PR modifies the UI (e.g. new buttons, new UI components, changing the padding/spacing/sizing, moving components, etc) or modifies the form input styles:
    - [x] I verified that all the inputs inside a form are aligned with each other.
    - [x] I added `Design` label and/or tagged `@Expensify/design` so the design team can review the changes.
- [x] If a new page is added, I verified it's using the `ScrollView` component to make it scrollable when more elements are added to the page.
- [x] If the `main` branch was merged into this PR after a review, I tested again and verified the outcome was still expected according to the `Test` steps.

### Screenshots/Videos
<details>
<summary>Android: Native</summary>

</details>

<details>
<summary>Android: mWeb Chrome</summary>

</details>

<details>
<summary>iOS: Native</summary>

</details>

<details>
<summary>iOS: mWeb Safari</summary>

</details>

<details>
<summary>MacOS: Chrome / Safari</summary>

https://github.com/user-attachments/assets/7c4c06c9-58c4-4e32-b993-d90e1dc966a8

</details>

<details>
<summary>MacOS: Desktop</summary>

https://github.com/user-attachments/assets/45811cfe-f081-4220-9c19-7d673fd470b5

</details>

### CODE REVIEW

**Comment by hungvu193** [comment]
Note: It looks like BE currently doesn't throw errors with invalid validate code.

**Review by situchan — CHANGES_REQUESTED** [review]
image

**Comment by hungvu193** [comment]
> image

I see. This is not related to current changes, however I think it's easy to update it so I can include useOnyx migration in this PR

**Comment by hungvu193** [comment]
Onyx migrating 's done. I mark this PR is ready so the linked issue won't be overdue.

**Comment by situchan** [comment]
## Reviewer Checklist

- [ ] I have verified the author checklist is complete (all boxes are checked off).
- [ ] I verified the correct issue is linked in the `### Fixed Issues` section above
- [ ] I verified testing steps are clear and they cover the changes made in this PR
    - [ ] I verified the steps for local testing are in the `Tests` section
    - [ ] I verified the steps for Staging and/or Production testing are in the `QA steps` section
    - [ ] I verified the steps cover any possible failure scenarios (i.e. verify an input displays the correct error message if the entered data is not correct)
    - [ ] I turned off my network connection and tested it while offline to ensure it matches the expected behavior (i.e. verify the default avatar icon is displayed if app is offline)
- [ ] I checked that screenshots or videos are included for tests on [all platforms](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#make-sure-you-can-test-on-all-platforms)
- [ ] I included screenshots or videos for tests on [all platforms](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#make-sure-you-can-test-on-all-platforms)
- [ ] I verified tests pass on **all platforms** & I tested again on:
    - [ ] Android: Native
    - [ ] Android: mWeb Chrome
    - [ ] iOS: Native
    - [ ] iOS: mWeb Safari
    - [ ] MacOS: Chrome / Safari
    - [ ] MacOS: Desktop
- [ ] If there are any errors in the console that are unrelated to this PR, I either fixed them (preferred) or linked to where I reported them in Slack
- [ ] I verified proper code patterns were followed (see [Reviewing the code](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md#reviewing-the-code))
    - [ ] I verified that any callback methods that were added or modified are named for what the method does and never what callback they handle (i.e. `toggleReport` and not `onIconClick`).
    - [ ] I verified that comments were added to code that is not self explanatory
    - [ ] I verified that any new or modified comments were clear, correct English, and explained "why" the code was doing something instead of only explaining "what" the code was doing.
    - [ ] I verified any copy / text shown in the product is localized by adding it to `src/languages/*` files and using the [translation method](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60)
    - [ ] I verified all numbers, amounts, dates and phone numbers shown in the product are using the [localization methods](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60-L68)
    - [ ] I verified any copy / text that was added to the app is grammatically correct in English. It adheres to proper capitalization guidelines (note: only the first word of header/labels should be capitalized), and is either coming verbatim from figma or has been approved by marketing (in order to get marketing approval, ask the Bug Zero team member to add the Waiting for copy label to the issue)
    - [ ] I verified proper file naming conventions were followed for any new files or renamed files. All non-platform specific files are named after what they export and are not named "index.js". All platform-specific files are named for the platform the code supports as outlined in the README.
    - [ ] I verified the JSDocs style guidelines (in [`STYLE.md`](https://github.com/Expensify/App/blob/main/contributingGuides/STYLE.md#jsdocs)) were followed
- [ ] If a new code pattern is added I verified it was agreed to be used by multiple Expensify engineers
- [ ] I verified that this PR follows the guidelines as stated in the [Review Guidelines](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md)
- [ ] I verified other components that can be impacted by these changes have been tested, and I retested again (i.e. if the PR modifies a shared library or component like `Avatar`, I verified the components using `Avatar` have been tested & I retested again)
- [ ] I verified all code is DRY (the PR doesn't include any logic written more than once, with the exception of tests)
- [ ] I verified any variables that can be defined as constants (ie. in CONST.js or at the top of the file that uses the constant) are defined as such
- [ ] If a new component is created I verified that:
    - [ ] A similar component doesn't exist in the codebase
    - [ ] All props are defined accurately and each prop has a `/** comment above it */`
    - [ ] The file is named correctly
    - [ ] The component has a clear name that is non-ambiguous and the purpose of the component can be inferred from the name alone
    - [ ] The only data being stored in the state is data necessary for rendering and nothing else
    - [ ] For Class Components, any internal methods passed to components event handlers are bound to `this` properly so there are no scoping issues (i.e. for `onClick={this.submit}` the method `this.submit` should be bound to `this` in the constructor)
    - [ ] Any internal methods bound to `this` are necessary to be bound (i.e. avoid `this.submit = this.submit.bind(this);` if `this.submit` is never passed to a component event handler like `onClick`)
    - [ ] All JSX used for rendering exists in the render method
    - [ ] The component has the minimum amount of code necessary for its purpose, and it is broken down into smaller components in order to separate concerns and functions
- [ ] If any new file was added I verified that:
    - [ ] The file has a description of what it does and/or why is needed at the top of the file if the code is not self explanatory
- [ ] If a new CSS style is added I verified that:
    - [ ] A similar style doesn't already exist
    - [ ] The style can't be created with an existing [StyleUtils](https://github.com/Expensify/App/blob/main/src/utils/index.ts) function (i.e. `StyleUtils.getBackgroundAndBorderStyle(theme.componentBG`)
- [ ] If the PR modifies code that runs when editing or sending messages, I tested and verified there is no unexpected behavior for all supported markdown - URLs, single line code, code blocks, quotes, headings, bold, strikethrough, and italic.
- [ ] If the PR modifies a generic component, I tested and verified that those changes do not break usages of that component in the rest of the App (i.e. if a shared library or component like `Avatar` is modified, I verified that `Avatar` is working as expected in all cases)
- [ ] If the PR modifies a component related to any of the existing Storybook stories, I tested and verified all stories for that component are still working as expected.
- [ ] If the PR modifies a component or page that can be accessed by a direct deeplink, I verified that the code functions as expected when the deeplink is used - from a logged in and logged out account.
- [ ] If the PR modifies the UI (e.g. new buttons, new UI components, changing the padding/spacing/sizing, moving components, etc) or modifies the form input styles:
    - [ ] I verified that all the inputs inside a form are aligned with each other.
    - [ ] I added `Design` label and/or tagged `@Expensify/design` so the design team can review the changes.
- [ ] If a new page is added, I verified it's using the `ScrollView` component to make it scrollable when more elements are added to the page.
- [ ] If the `main` branch was merged into this PR after a review, I tested again and verified the outcome was still expected according to the `Test` steps.
- [ ] I have checked off every checkbox in the PR reviewer checklist, including those that don't apply to this PR.

### Screenshots/Videos
<details>
<summary>Android: Native</summary>



</details>

<details>
<summary>Android: mWeb Chrome</summary>



</details>

<details>
<summary>iOS: Native</summary>



</details>

<details>
<summary>iOS: mWeb Safari</summary>



</details>

<details>
<summary>MacOS: Chrome / Safari</summary>



</details>

<details>
<summary>MacOS: Desktop</summary>



</details>

**Review by NikkiWines — CHANGES_REQUESTED** [review]

**Comment by hungvu193** [comment]
@NikkiWines What's the successful response? Can you give me the example? I'm seeing here's the error response, it merges `requestPhysicalCardForm` key.

Screenshot 2024-11-01 at 11 45 15

**Comment by hungvu193** [comment]
Another thing, The error that's showing between assigned card sections and inside ValidateCodeActionModal are different, do we need to fix it as well?


https://github.com/user-attachments/assets/770123a8-3d71-43d8-8d90-3191b1d4c7c7

**Comment by hungvu193** [comment]
I see. Looks like the isSucessful variable was calculated incorrectly. I'll take a look

**Comment by hungvu193** [comment]
I'm a bit confused about the way we get selected card to issue here:
https://github.com/Expensify/App/blob/bef062b4caa7f665159dc107911e708031e648c4/src/pages/settings/Wallet/Card/BaseGetPhysicalCard.tsx#L113-L114

Once the successful data arrives, `cardToBeIssued` will become undefined because its state is now `NOT_ACTIVATED` (4).

**Comment by NikkiWines** [comment]
> Once the successful data arrives, cardToBeIssued will become undefined because its state is now NOT_ACTIVATED (4).

Yeah, I think that's expected, though - the card is no longer not issued; it's been shipped / queued to be shipped to the user, so it's in a non-activated state until they enter the CVV on the back of the card to verify it's been received.

**Comment by situchan** [comment]
Please reassign another C+ as I'll be OOO

**Comment by hungvu193** [comment]
> > Once the successful data arrives, cardToBeIssued will become undefined because its state is now NOT_ACTIVATED (4).
> 
> Yeah, I think that's expected, though - the card is no longer not issued; it's been shipped / queued to be shipped to the user, so it's in a non-activated state until they enter the CVV on the back of the card to verify it's been received.

Yeah, but it also means `cardToBeIssued.isSuccessful` will become undefined, so it won't navigate to the next step. I'm finding another way to detect when card is issued sucessfully.

**Comment by hungvu193** [comment]
@NikkiWines Can you check again? I updated the new condition to navigate user to the next route when API was called successfully.

**Comment by NikkiWines** [comment]
It works! But I might double-check how we handle the success of validateCode flow for other routes, the end feels a little jarring/abrupt? Maybe a success page or something. 

@mountiny do you have any idea? Do we typically show a success message / success page or something for this? 

https://github.com/user-attachments/assets/b0081b0a-d2f2-4866-bd06-43f5aa152576

**Comment by mountiny** [comment]
@NikkiWines That looks fine to me personally, but I agree we could add a success page. I think it does not have to be in scope of this PR. @koko57 added something similar to when domain is provisioned. We could create issue for the design team to whip something nice up

**Comment by NikkiWines** [comment]
Yeah it might be a good thing to add to that re-usable component in general, but agreed it's not a blocker for this PR or necessarily needs to be lumped in to this issue specifically. cc: @Expensify/design thoughts on adding some sort of success notification or page to the newDot in app validation flow? (see [here](https://github.com/Expensify/App/pull/51135#issuecomment-2460806804))

**Comment by NikkiWines** [comment]
@hungvu193 is there anything you need to move forward here? Is this ready for review by @dominictb? (thanks for stepping in @dominictb!)

**Comment by dannymcclain** [comment]
Definitely not against a success/confirmation page—we have lots of them sprinkled throughout the app so that shouldn't be a problem. (Also fine with it being a follow-up/not strictly part of this PR).

Also, is it possible to update this screen to match our other magic code screens? The `Verify` button should be at the bottom like our normal "footer buttons" and any errors should show above it.

CleanShot 2024-11-07 at 15 46 52@2x

cc @Expensify/design

**Comment by hungvu193** [comment]
> @hungvu193 is there anything you need to move forward here? Is this ready for review by @dominictb? (thanks for stepping in @dominictb!)

It's ready for review

**Comment by hungvu193** [comment]
> Also, is it possible to update this screen to match our other magic code screens? The Verify button should be at the bottom like our normal "footer buttons" and any errors should show above it.
> 
> 

Ah yeah, we had another PR to fix it, it was merged already, I need to merge this PR with main then the button will align bottom

**Review by dominictb — CHANGES_REQUESTED** [review]

File: src/pages/settings/Wallet/Card/BaseGetPhysicalCard.tsx [hunk 1]
```diff
@@ -144,19 +136,37 @@ function BaseGetPhysicalCard({
...
+            setActionCodeModalVisible(true);
             return;
         }
         GetPhysicalCardUtils.goToNextPhysicalCardRoute(domain, updatedPrivatePersonalDetails);
-    }, [cardID, cardToBeIssued?.cardID, domain, draftValues, isConfirmation, session?.authToken, privatePersonalDetails]);
+    }, [isConfirmation, domain, draftValues, privatePersonalDetails]);
+
+    const handleIssuePhysicalCard = useCallback(
+        (validateCode: string) => {
+            setCurrentCardId(cardToBeIssued?.cardID.toString());
```
[the remark this answers is not in the document]
Comment (hungvu193) [reply src/pages/settings/Wallet/Card/BaseGetPhysicalCard.tsx:162]: I explained it [here](https://github.com/Expensify/App/pull/51135#issuecomment-2456617331). `cardToBeIssued` will become undefined right after Pusher event arrives.

File: src/libs/actions/Wallet.ts [hunk 2]
```diff
@@ -288,15 +291,78 @@ function requestPhysicalExpensifyCard(cardID: number, authToken: string, private
...
 function resetWalletAdditionalDetailsDraft() {
     Onyx.set(ONYXKEYS.FORMS.WALLET_ADDITIONAL_DETAILS_DRAFT, null);
 }
 
+/**
+ * Clear the error of specific card
+ * @param cardId The card id of the card that you want to clear the errors.
+ */
+function clearPhysicalCardError(cardId: string) {
+    FormActions.clearErrors(ONYXKEYS.FORMS.REPORT_PHYSICAL_CARD_FORM);
```
[the remark this answers is not in the document]
Comment (hungvu193) [reply src/libs/actions/Wallet.ts:358]: Yep. We might need it. There's case where user pressed `Close` button right next to the error message. (`clearError` props of `ValidateCodeActionModal`)

File: src/pages/settings/Wallet/Card/BaseGetPhysicalCard.tsx (hunk 1 above)
[the remark this answers is not in the document]
Comment (hungvu193) [reply src/pages/settings/Wallet/Card/BaseGetPhysicalCard.tsx]: No, `currentCardID` will be initial once on the very first render. So there won't be an issue

**Review by dominictb — CHANGES_REQUESTED** [review]

**Comment by hungvu193** [comment]
> @hungvu193 Are [#51135 (comment)](https://github.com/Expensify/App/pull/51135#issuecomment-2424826532) and [#51135 (comment)](https://github.com/Expensify/App/pull/51135#issuecomment-2451425006) expected? Or are we waiting for BE changes?

I think they're expected for now. I believe we don't want to block existing user from using this feature at the moment because we haven't released this PR yet. cc @NikkiWines

**Comment by NikkiWines** [comment]
@dominictb you might be able to test more easily by following the setup steps from [this PR](https://github.com/Expensify/App/pull/38998)? @situchan it might be helpful to incorporate something like this into the QA steps as well

File: src/components/ValidateCodeActionModal/ValidateCodeForm/BaseValidateCodeForm.tsx [hunk 3]
```diff
@@ -65,6 +65,9 @@ type ValidateCodeFormProps = {
 
     /** Function is called when validate code modal is mounted and on magic code resend */
     sendValidateCode: () => void;
+
+    /** Wheather the form is loading or not */
```
Comment (NikkiWines) [inline src/components/ValidateCodeActionModal/ValidateCodeForm/BaseValidateCodeForm.tsx:69]:
```suggestion
    /** Whether the form is loading*/
```

File: src/components/ValidateCodeActionModal/type.ts [hunk 4]
```diff
@@ -37,6 +37,9 @@ type ValidateCodeActionModalProps = {
 
     /** If the magic code has been resent previously */
     hasMagicCodeBeenSent?: boolean;
+
+    /** Wheather the form is loading or not */
```
Comment (NikkiWines) [inline src/components/ValidateCodeActionModal/type.ts:41]:
```suggestion
    /** Whether the form is loading */
```

File: src/libs/actions/Wallet.ts [hunk 5]
```diff
@@ -288,15 +291,78 @@ function requestPhysicalExpensifyCard(cardID: number, authToken: string, private
...
 
 function resetWalletAdditionalDetailsDraft() {
     Onyx.set(ONYXKEYS.FORMS.WALLET_ADDITIONAL_DETAILS_DRAFT, null);
 }
 
+/**
+ * Clear the error of specific card
+ * @param cardId The card id of the card that you want to clear the errors.
+ */
+function clearPhysicalCardError(cardId: string) {
```
Comment (NikkiWines) [inline src/libs/actions/Wallet.ts:357]:
```suggestion
function clearPhysicalCardError(cardID: string) {
```

File: src/pages/settings/Wallet/Card/BaseGetPhysicalCard.tsx [hunk 6]
```diff
@@ -144,19 +137,38 @@ function BaseGetPhysicalCard({
         isRouteSet.current = true;
     }, [cardList, currentRoute, domain, domainCards.length, draftValues, loginList, cardToBeIssued, privatePersonalDetails]);
 
+    useEffect(() => {
+        // Current step of the get physical card flow should be the confirmation page; and
+        // Card has NOT_ACTIVATED state when successfully being issued so cardToBeIssued should be undefined
+        if (!isConfirmation || !!cardToBeIssued || !currentCardID) {
+            return;
+        }
+        // Form draft data needs to be erased when the flow is complete,
```
Comment (NikkiWines) [inline src/pages/settings/Wallet/Card/BaseGetPhysicalCard.tsx:146]:
```suggestion

        // Form draft data needs to be erased when the flow is complete,
```

**Review by NikkiWines — CHANGES_REQUESTED** [review]
Couple of comments

**Comment by hungvu193** [comment]
> Couple of comments

Addressed all the comments

File: src/pages/settings/Wallet/Card/BaseGetPhysicalCard.tsx
[the remark this answers is not in the document]
Comment (hungvu193) [reply src/pages/settings/Wallet/Card/BaseGetPhysicalCard.tsx:1]: Oh shit. Missed thiss one, gonna take a look today

**Comment by NikkiWines** [comment]
@hungvu193 you've got conflicts

File: src/pages/settings/Wallet/Card/BaseGetPhysicalCard.tsx
[the remark this answers is not in the document]
Comment (hungvu193) [reply src/pages/settings/Wallet/Card/BaseGetPhysicalCard.tsx:1]: Fixed. Can you give it a retest? ty

**Comment by NikkiWines** [comment]
@dominictb can you re-review please 🙇

File: src/pages/settings/Wallet/Card/BaseGetPhysicalCard.tsx
[the remark this answers is not in the document]
Comment (hungvu193) [reply src/pages/settings/Wallet/Card/BaseGetPhysicalCard.tsx:1]: It looks like we decided to add the default value of IDs inside the App to `-1`.
This [huge PR](https://github.com/Expensify/App/pull/42634) did it. I thought we only apply `-1` to the default value of `reportID` but I think I was wrong 😄

**Review by dominictb — CHANGES_REQUESTED** [review]

File: src/pages/settings/Wallet/Card/BaseGetPhysicalCard.tsx [hunk 7]
```diff
@@ -144,19 +137,40 @@ function BaseGetPhysicalCard({
...
         }
         GetPhysicalCardUtils.goToNextPhysicalCardRoute(domain, updatedPrivatePersonalDetails);
-    }, [cardID, cardToBeIssued?.cardID, domain, draftValues, isConfirmation, session?.authToken, privatePersonalDetails]);
+    }, [isConfirmation, domain, draftValues, privatePersonalDetails]);
+
+    const handleIssuePhysicalCard = useCallback(
+        (validateCode: string) => {
+            setCurrentCardID(cardToBeIssued?.cardID.toString());
+            const updatedPrivatePersonalDetails = GetPhysicalCardUtils.getUpdatedPrivatePersonalDetails(draftValues, privatePersonalDetails);
+            Wallet.requestPhysicalExpensifyCard(cardToBeIssued?.cardID ?? -1, session?.authToken ?? '', updatedPrivatePersonalDetails, validateCode);
```
[the remark this answers is not in the document]
Comment (hungvu193) [reply src/pages/settings/Wallet/Card/BaseGetPhysicalCard.tsx:169]: Sure. I'll do it shortly

**Comment by hungvu193** [comment]
Yes, It's to prevent excessive API calls. But I think we should also fix it, we should add optimistic data to our API that use `validateCode` param.

File: src/pages/settings/Wallet/Card/BaseGetPhysicalCard.tsx (hunk 7 above)
[the remark this answers is not in the document]
Comment (hungvu193) [reply src/pages/settings/Wallet/Card/BaseGetPhysicalCard.tsx:169]: We clear the error when modal is hide, unfortunately we used the old cardID value for it, which caused the issue, can you verify it now? Ty

**Comment by NikkiWines** [comment]
@hungvu193 what's the latest here, can @dominictb re-review now?

**Comment by hungvu193** [comment]
> @hungvu193 what's the latest here, can @dominictb re-review now?

Yes. I commented [here](https://github.com/Expensify/App/pull/51135#discussion_r1843417718)

**Review by dominictb — CHANGES_REQUESTED** [review]

**Comment by hungvu193** [comment]
Merged main and addressed all your comments 😄

**Review by dominictb — APPROVED** [review]
All yours @NikkiWines

**Review by NikkiWines — COMMENTED** [review]
Looks like this needs another web PR to fix the error handling. I've got a PR up for that now and I'll CP if we can CP it so it goes live tomorrow. Holding off on approving until then.

Flow looks good once that correct error is being returned 🪅 
https://github.com/user-attachments/assets/f99ab563-0978-4d94-9c8e-a87671eb225e


However, it looks like the error for the `Assigned cards` needs to be adjusted here so it's not just displaying `Oops!...` - maybe we can display the same error (or just red dot?) in this case? 

image

**Comment by NikkiWines** [comment]
Backend PR for the error handling is on staging, will be deployed tomorrow 😊

**Comment by hungvu193** [comment]
> However, it looks like the error for the `Assigned cards` needs to be adjusted here so it's not just displaying `Oops!...` - maybe we can display the same error (or just red dot?) in this case?

Yeah, currently we're getting the error from `ONYXKEYS.CARD_LIST` for assigned card, but when we validate the magic code the errors sometimes return from `[ONYXKEYS.FORMS.REPORT_PHYSICAL_CARD_FORM](https://github.com/Expensify/App/pull/51135#issuecomment-2451285601)`.

**Comment by NikkiWines** [comment]
Backend logic for the error handling is on prod

**Comment by NikkiWines** [comment]
No - it just handles updating the error message that we return  when getting a 401 from the  `RequestPhysicalExpensifyCard`command. 

However, I think this is probably ok for now. I'd like to not display anything other than the 🔴  in that scenario, rather than show the `Oops!` but it's a minor thing. 

I'm going to double check to see if we have staging friendly QA steps for this command, because I don't think we want go through the real flow of a requesting a new card in this case - once i've got that settled I'll re-review and merge.

**Review by NikkiWines — APPROVED** [review]
:shipit:

**Comment by NikkiWines** [comment]
Confirmed this works as expected on staging: 

https://github.com/user-attachments/assets/f9b039a9-a985-467c-b70a-e7f30639a178

**Pull request merged**